cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,489 @@
+ * to be copied to the output. Length must be equal to `num_destination_rows`
+ */
+template <bool left_have_valids, bool right_have_valids>
+__global__ void materialize_merged_bitmask_kernel(cudf::column_device_view left_dcol,
+                                                  cudf::column_device_view right_dcol,
+                                                  cudf::mutable_column_device_view out_dcol,
+                                                  cudf::size_type const num_destination_rows,
+                                                  index_type const* const __restrict__ merged_indices) {
+  cudf::size_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+
+  cudf::bitmask_type const* const __restrict__ source_left_mask = left_dcol.null_mask();
+  cudf::bitmask_type const* const __restrict__ source_right_mask= right_dcol.null_mask();
+  cudf::bitmask_type* const __restrict__ destination_mask = out_dcol.null_mask();
+  
+  auto active_threads =
+    __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  while (destination_row < num_destination_rows) {
+    index_type const& merged_idx = merged_indices[destination_row];
+    side const src_side = thrust::get<0>(merged_idx);
+    cudf::size_type const src_row  = thrust::get<1>(merged_idx);
+    bool const from_left{src_side == side::LEFT};
+    bool source_bit_is_valid{true};
+    if (left_have_valids && from_left) {
+      source_bit_is_valid = left_dcol.is_valid_nocheck(src_row);
+    }
+    else if (right_have_valids && !from_left) {
+      source_bit_is_valid = right_dcol.is_valid_nocheck(src_row);
+    }
+
+    // Use ballot to find all valid bits in this warp and create the output
+    // bitmask element
+    cudf::bitmask_type const result_mask{
+      __ballot_sync(active_threads, source_bit_is_valid)};
+
+    cudf::size_type const output_element = cudf::word_index(destination_row);
+
+    // Only one thread writes output
+    if (0 == threadIdx.x % warpSize) {
+      destination_mask[output_element] = result_mask;
+    }
+
+    destination_row += blockDim.x * gridDim.x;
+    active_threads =
+      __ballot_sync(active_threads, destination_row < num_destination_rows);
+  }
+}
+
+void materialize_bitmask(cudf::column_view const& left_col,
+                         cudf::column_view const& right_col,
+                         cudf::mutable_column_view& out_col,
+                         index_type const* merged_indices,
+                         cudaStream_t stream) {
+  constexpr cudf::size_type BLOCK_SIZE{256};
+  cudf::experimental::detail::grid_1d grid_config {out_col.size(), BLOCK_SIZE };
+
+  auto p_left_dcol  = cudf::column_device_view::create(left_col);
+  auto p_right_dcol = cudf::column_device_view::create(right_col);
+  auto p_out_dcol   = cudf::mutable_column_device_view::create(out_col);
+
+  auto left_valid  = *p_left_dcol;
+  auto right_valid = *p_right_dcol;
+  auto out_valid   = *p_out_dcol;
+
+  //these tests in the legacy code
```

Comment:
Shouldn't need to refer to the "legacy code". If the old code was wrong, just fix it. Then delete this long comment.

Reply:
Done.